File: java/simulation/field/grassField/GrassNode.java
package simulation.field.grassField;

import java.awt.Color;

import simulation.common.Node;

public class GrassNode extends Node {

private static final int RED = 149;
private static final int GREEN = 250;
private static final int  BLUE = 235;

static final int maxGrassHeight = 10;
private static final int growthIntervalRed = RED/maxGrassHeight;
private static final int growthIntervalGreen = GREEN/maxGrassHeight;

//transient used to stop attribute from being serialized
public transient Color color;
private transient  Color cut;

public GrassNode(int x, int y, int grassHeight){
	object = "grass";
	//unCut = new Color(149, 0, 235);
	cut = new Color(0, GREEN, BLUE);
	this.x = x;
	this.y = y;
	
	this.attribute = grassHeight;
	color = new Color(grassHeight * growthIntervalRed, 255 - (grassHeight*growthIntervalGreen), BLUE);
}

public void growGrass(){
	if(attribute < maxGrassHeight){
		attribute++;
	
		int greenValue = color.getGreen() - growthIntervalGreen;
		int redValue = color.getRed() + growthIntervalRed;
		
		if(greenValue < 0){
			greenValue = 0;
		}
		
		if(redValue > RED){
			redValue = RED;
		}
		
		color = new Color(redValue, greenValue, color.getBlue());
		//System.out.println("grass has been grown " + toString());
	}
}

public void cutGrass(){
	attribute = 0;
	color = cut;
}

public int getGrassHeight(){
	return attribute;
}

public String toString(){
	return x + " " + y;
}

public void setBoundaryColor(){
	color = color.darker().darker().darker();
}
}

File: src/simulation/field/grassField/GrassNode.ts
import { Node } from '../../common/Node';

const MAX_GRASS_HEIGHT = 10;
const RED = 149;
const GREEN = 250;
const BLUE = 235;
const GROWTH_INTERVAL_RED = Math.floor(RED / MAX_GRASS_HEIGHT);
const GROWTH_INTERVAL_GREEN = Math.floor(GREEN / MAX_GRASS_HEIGHT);

export class GrassNode extends Node {
  static readonly maxGrassHeight: number = MAX_GRASS_HEIGHT;

  constructor(x: number, y: number, grassHeight: number) {
    super();
    this.object = 'grass';
    this.x = x;
    this.y = y;
    this.attribute = grassHeight;
  }

  growGrass(): void {
    if (this.attribute < MAX_GRASS_HEIGHT) {
      this.attribute++;
    }
  }

  cutGrass(): void {
    this.attribute = 0;
  }

  getGrassHeight(): number {
    return this.attribute;
  }

  toString(): string {
    return this.x + ' ' + this.y;
  }

  // no-op: colour is not used in headless simulation
  setBoundaryColor(): void {}
}

export { RED, GREEN, BLUE, GROWTH_INTERVAL_RED, GROWTH_INTERVAL_GREEN };
